### PR TITLE
Cleanup Marshallable, changes to addVertex implementation, introduced Atom, sbt root project, and README

### PR DESCRIPTION
**Pull request by dkrieg**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -55,9 +55,12 @@ val commonSettings = Seq(
       </developers>
 )
 
+lazy val root = project.in(file("."))
+  .aggregate(`gremlin-scala`, macros)
+
```
Comment (joan38) [inline build.sbt]: Is this useful?
Please read "Default root project" :
http://www.scala-sbt.org/0.13/tutorial/Multi-Project.html

Comment (dkrieg) [inline build.sbt]: Though the build works fine without the declared root project I do find utility in it for with it, when you are in an `sbt` session and type `projects` the following information is output:
```
> projects
[info] In file:/Projects/gremlin-scala/
[info]     gremlin-scala
[info]     macros
[info]   * root
> 
```
Without it the following is output:
```
> projects
[info] In file:/Projects/gremlin-scala/
[info]   * default-28afd7
[info]     gremlin-scala
[info]     macros
>
```

File: gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala [hunk 2]
```diff
@@ -6,21 +6,25 @@ import shapeless._
...
+      case (k, v) =>
+      val label = T.label.name
+      val id = T.id.name
+      k match {
+        case `label` => Seq(T.label, v.asInstanceOf[AnyRef])
+        case `id` => Seq(T.id, v.asInstanceOf[AnyRef])
+        case _ => Seq(k, v.asInstanceOf[AnyRef])
+      }
+    }: _*)
   }
```
Comment (joan38) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala]: Please see http://milessabin.com/blog/2011/06/09/scala-union-types-curry-howard/
And use:

    def addVertex(properties: Map[(T |∨| String)#λ, Any]): ScalaVertex = ...

File: gremlin-scala/src/main/scala/gremlin/scala/ScalaVertex.scala [hunk 3]
```diff
@@ -53,9 +53,16 @@ case class ScalaVertex(vertex: Vertex) extends ScalaElement[Vertex] {
...
+      case (k, v) =>
+        val label = T.label.name
+        val id = T.id.name
+        k match {
+          case `label` => Seq(T.label, v.asInstanceOf[AnyRef])
+          case `id` => Seq(T.id, v.asInstanceOf[AnyRef])
+          case _ => Seq(k, v.asInstanceOf[AnyRef])
+        }
+    }: _*)
   }
```
Comment (joan38) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaVertex.scala]: Same as up

File: build.sbt (hunk 1 above)
Comment (joan38) [inline build.sbt]: Ok. As you like

File: gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala (hunk 2 above)
Comment (dkrieg) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala]: BRILLIANT!  Thanks for the tip, this is exactly right!

Comment (dkrieg) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala]: The syntax to make this work is actually:
```
import org.apache.tinkerpop.gremlin.structure.T
import shapeless.|∨|

def addVertex[K: (T |∨| String)#λ](p: Map[K, Any]) = p

addVertex(Map(T.label -> "test", T.id -> 9, "name" -> "value"))
```
which does not actually run.  From `console` we get:
```
scala> addVertex(Map(T.label -> "test", T.id -> 9, "name" -> "value"))
<console>:24: error: Cannot prove that (Comparable[_ >: String with org.apache.tinkerpop.gremlin.structure.T <: Comparable[_ >: String with org.apache.tinkerpop.gremlin.structure.T <: java.io.Serializable] with java.io.Serializable] with java.io.Serializable => Nothing) => Nothing <:< org.apache.tinkerpop.gremlin.structure.T => Nothing with String => Nothing => Nothing.
       addVertex(Map(T.label -> "test", T.id -> 9, "name" -> "value"))
 ```
Any advise here?

Comment (dkrieg) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala]: What about something like this?
```
import org.apache.tinkerpop.gremlin.structure.T
import shapeless.HMap

class P[K, V]
implicit def tToAny[V <: Any] = new P[T, V]
implicit def stringToAny[V <: Any] = new P[String, V]

def addVertex(p: HMap[P]) = p

addVertex(HMap[P](T.label -> "test", T.id -> 9, "name" -> "value", "weight" -> 9.4))
```

Comment (dkrieg) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala]: After spending much time - I cannot get this to work as desired.

Comment (dkrieg) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala]: Could do:
```scala
import org.apache.tinkerpop.gremlin.structure.T

def addVertex(p: (Either[T, String], Any)*) = p.foldLeft(List[Object]()) {
  import org.apache.tinkerpop.gremlin.structure.T._
  (xs, x) => (x._1 match {
    case Left(`label`) => label.asInstanceOf[Object]
    case Left(`id`) => id.asInstanceOf[Object]
    case Right(k) => k.asInstanceOf[Object]
  }) :: x._2.asInstanceOf[Object] :: xs
}

addVertex(Left(T.label) -> "testlabel", Left(T.id) -> 0, Right("name") -> "marko", Right("age") -> 40)
```

File: gremlin-scala/src/main/scala/gremlin/scala/ScalaVertex.scala (hunk 3 above)
Comment (dkrieg) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaVertex.scala]: @joan38 Same as up.

File: gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala (hunk 2 above)
Comment (joan38) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala]: That's not very nice can we use the HMap from shapeless? And therefore will not be able to use the varargs :/

Comment (dkrieg) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala]: HMap does not  support the map and flatMap functions needed to convert into the varargs required in the Java API.

Comment (dkrieg) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala]: What about this:
```scala
  def addVertex(label: (T, Any), properties: (String, Any)*): ScalaVertex = {
    graph.addVertex(properties.foldLeft(List[Object](label._1, label._2.asInstanceOf[Object])) {
      (xs, x) => x._1.asInstanceOf[Object] :: x._2.asInstanceOf[Object] :: xs
    }: _*)
  }
```

Comment (dkrieg) [inline gremlin-scala/src/main/scala/gremlin/scala/ScalaGraph.scala]: Seems we are back to
```scala
def addVertex(label: String, properties: (String, Any)*): ScalaVertex = {
  graph.addVertex(label).setProperties(properties.toMap)
}
```

**Comment by dkrieg** [comment]
If you are not interested in this PR I will cancel it.

**Comment by mpollmeier** [comment]
Sorry about the misunderstanding, I was waiting for your final "it's ready now" comment.. in chat you mentioned you still needed to fix something. I'll have a look when I have a minute.

**Pull request merged**